python/src/opendp/extras/polars/__init__.py

```diff
@@ -864,6 +866,45 @@ def with_columns_seq(  # type: ignore[empty-body]
             """
             ...
 
+        def join(  # type: ignore[empty-body]
+            self,
+            other: _LazyFrame,
+            on: str | _Expr | Sequence[str | _Expr] | None = None,
+            how: JoinStrategy = "inner",
+            *,
+            left_on: str | _Expr | Sequence[str | _Expr] | None = None,
+            right_on: str | _Expr | Sequence[str | _Expr] | None = None,
+            suffix: str = "_right",
+            validate: JoinValidation = "m:m",
+            join_nulls: bool = False,
+            coalesce: bool | None = None,
+            allow_parallel: bool = True,
+            force_parallel: bool = False,
+        ) -> LazyFrameQuery:
+            """
+            Add a join operation to the Logical Plan.
+            """
+            ...
+
+        def with_keys(
+            self,
+            keys: _LazyFrame,
+            on: list[str] | None = None,
+        ) -> LazyFrameQuery:
+            """
+            Shorthand to join with an explicit key-set.
+            """
```

Comment:
`with_keys` is in the notebook, once. Consider using the full version with `join` first, and then demonstrating the shortcut?

Reply:
... but unless you think this is really a big improvement, I wonder if it would be better just not to add shortcuts that diverge farther from the out-of-the-box polars syntax? Constraining users to a polars-like syntax may not be a bad thing.

Reply:
I was curious how you'd feel about this! Some considerations:
1. Writing a left join is more difficult in the context API: see the complexity of this implementation, where you have to go under the hood. This gives an easier shorthand to write a left join.
2. Left joins are more likely to be supported by database backends.
3. I wanted this to be easy to use... with the Polars API the key set needs to be lazy, the user has to specify they want a right join and they need to specify join keys. 

Added an example with explicit join to the notebook.

Reply:
Sounds good. Maybe record these motivations in a comment?

Reply:
Added